Create Rancher Component Library

rak-phillip opened the pull request:
This is the first of a series of PRs for porting a collection of components into a new project so that they can be published as a component framework for use in other Rancher user interfaces. My goal for this initial step is to keep our components as unchanged as possible to limit the surface area for regressions after we start using this library in our apps.  Given the size of this PR, I think it makes sense to roll this out in steps:

1. Create a new package to house a component framework (this PR)
1. Set up github actions for releasing and publishing Rancher Components
1. Install Rancher Components as a dependency for Rancher Dashboard and remove duplicated components

This PR is minimal in what it covers, but I hope that it will serve as a basis for getting started with a component library that we can share across projects.

Still to do:

* App CSS. Components are bundled with their own styles, which can contribute to bloat. I would like to isolate component styles from the rest of Rancher Dashboard.
* Documentation. Minimally, we should be able to introduce Rancher Components and provide a quick way to get started in the readme.
* Optimize bundle size. Right out of the gate, Rancher Components raises warnings about bundle size. We should make sure that we are tree-shaking and lazy loading where it makes sense.
* Port more components 🙂
* Add unit tests
* More typescript support
* Create a Utility Library

Contributes to rancher/dashboard#5108

rak-phillip commented on pkg/rancher-components/types/index.d.ts:
```diff
@@ -0,0 +1,11 @@
+import { VueConstructor } from 'vue'
+
+export const Badge: VueConstructor
```
Typing all of our components a `VueConstructor` seems to give us basic typescript support. I'm wondering if we could do better here.. I know that this is improved in Vue 3

cnotv replied:
There's an ongoing discussion to the methodology to be adopted for Typescript components and I'm updating the state in Confluence https://confluence.suse.com/pages/viewpage.action?pageId=969801913
We will have to discuss how to proceed for integrating TS in the components due the variety of cases.

rak-phillip commented on pkg/rancher-components/src/mixins/labeled-form-element.js:
```diff
@@ -0,0 +1,139 @@
+import $ from 'jquery'
```
I want to remove `jQuery` everywhere, but I think it's best to keep for now just so we aren't changing too much at once.

rak-phillip commented on pkg/rancher-components/.npmrc:
```diff
@@ -0,0 +1 @@
+@rancher:registry=https://npm.pkg.github.com
```
I think it makes sense to publish to the github repository, but we can use npm if that makes more sense.

rak-phillip replied:
We will use npm

rak-phillip commented on pkg/rancher-components/README.md:
````diff
@@ -0,0 +1,29 @@
+# rancher-components
+
+## Project setup
+```
+npm install
+```
+
+### Compiles and hot-reloads for development
+```
+npm run serve
+```
+
+### Compiles and minifies for production
+```
+npm run build
+```
+
+### Run your unit tests
+```
+npm run test:unit
+```
+
+### Lints and fixes files
+```
+npm run lint
+```
````
This should all be yarn.. update to include the build steps.

rak-phillip commented on pkg/rancher-components/src/components/BadgeState/BadgeState.spec.ts:
```diff
@@ -0,0 +1,14 @@
+import { shallowMount } from '@vue/test-utils'
```
I like the idea of keeping tests as close to the actual component implementation itself. I was thinking that we could do something similar with storybook as well if we think that would be beneficial. 

The tradeoff with the directory structure that I chose is that we kind of end up with a lot of ceremony around creating new components.. Vuetify does something similar.

cnotv replied:
The problem may be due the use of a flat tree file structure in the components, that may end up in a never ending scrolling for each time.
We may consider grouping components for keep files together.

cnotv commented on pkg/rancher-components/src/components/BadgeState/BadgeState.vue:
```diff
@@ -0,0 +1,100 @@
+<script>
```
If this does not slow you down, may we use Typescript for these first isolated components?

cnotv replied:
Ok forget it, I would rather avoid this till we setup correctly the project to support Typescript. I just tried to create a test for a TS component and this had some issues :cry:

rak-phillip replied:
I think this is a good idea and might be the next thing we could tackle. I don't anticipate there being any issue, but I'd like to keep the scope of this PR focused on standing up the component library and follow up with some long overdue refactoring in subsequent PRs.

rak-phillip replied:
Created #5932 so we can address this

cnotv replied:
@rak-phillip in the meanwhile the issue with TS has also been solved, but we can create tests afterwards without problems. I've created a bunch of tests cases if you are interested to take a look into: https://github.com/rancher/dashboard/pull/5779

richard-cox commented on shell/nuxt.config.js:
```diff
@@ -607,7 +607,7 @@ export default function(dir, _appConfig) {
     },
 
-    typescript: { typeCheck: { eslint: { files: './**/*.{ts,js,vue}' } } },
+    typescript: { typeCheck: { eslint: { files: './shell/**/*.{ts,js,vue}' } } },
```
This unfortunately means packages that should be checked, which i think covers all other cases, won't be covered. Can you check if there's an explicit way to exclude the component library

rak-phillip replied:
That was my concern with this change, but it wasn't clear to me that this behavior was intentional. I'll revert this change and focus some effort on excluding components.

richard-cox commented on pkg/rancher-components/public/index.html:
```diff
@@ -0,0 +1,17 @@
+<!DOCTYPE html>
```
What's the plan for these public assets? There's a lot of work going on for with storybook, Neil has also mentioned we should think about maybe a public site. My fear is that we would start duplicating effort if this became another visual guide to dashboard assets

cnotv replied:
We would need for sure 2 Storybooks: one as living styleguide for the component library, one for the UI of the dashboard, so we could verify the abstracted UI.
I am not so sure how much can be useful the second one, though.

rak-phillip replied:
I think this is a vestige of boilerplate and can be safely removed.. I would prefer that we don't duplicate these efforts as well 🙂

richard-cox reviewed (changes requested):
Duplicating `src/assets`
Interesting conundrum, though we shouldn't duplicate (_variables.scss has already started to diverge). I think in the short term we should import from outside the pkg structure ( _mixins.scss imports shell _mixins.scss) and discuss a styles pkg? 

Duplicating `src/components`
We shouldn't duplicate these either, keeping them in sync will be a pain and only get worse when more components are copied over. In the short term we should keep them in the pkg and remove from shell. This is the best time to start this process, early in the release cycle. In the future it should be pretty easy to just update references from pkg/rancher-components to @rancher/components.

Importing `utils`
I made a comment inline, but this could be the basis of another pkg for common code?

Exporting JS as TS
I don't think we need to worry about this yet. I _think_ we would only lose autocomplete of pkg'd components in external TS components? 
- We're not sure on how these components will look as well, even without the move to Vue3.
- Dropping this would remove the need for types/index.d.ts and <component>/index.ts files

Other points
- `cd pkg/rancher-components && yarn lint` fails, it's assuming untyped vue components are typed/ts
- ` cd pkg/rancher-components && yarn run build:lib` errors with lots of 
   ```
  Syntax Error: Thread Loader (Worker 1)
  Failed to load config "standard" to extend from.
  Referenced from: /home/richard/dev/github/rancher/dashboard-phil-ts/.eslintrc.js
      at _normalizeObjectConfigDataBody.next (<anonymous>)
      at _normalizeObjectConfigData.next (<anonymous>)
   ```

rak-phillip commented:
> Duplicating `src/assets` Interesting conundrum, though we shouldn't duplicate (_variables.scss has already started to diverge). I think in the short term we should import from outside the pkg structure ( _mixins.scss imports shell _mixins.scss) and discuss a styles pkg?

Much agreed, I was thinking that maybe we could define core application/component styles that we bundle for consumers of Rancher Components to import. 

> Duplicating `src/components` We shouldn't duplicate these either, keeping them in sync will be a pain and only get worse when more components are copied over. In the short term we should keep them in the pkg and remove from shell. This is the best time to start this process, early in the release cycle. In the future it should be pretty easy to just update references from pkg/rancher-components to @rancher/components.

This is correct, but we need to get to a point where we can publish a component library before we start replacing. This is the topic of #5935

> Importing `utils` I made a comment inline, but this could be the basis of another pkg for common code?

That is also correct, I mentioned this in the description and created a follow up issue so we can focus on resolving this. #5933

> Exporting JS as TS I don't think we need to worry about this yet. I _think_ we would only lose autocomplete of pkg'd components in external TS components?
> 
>     * We're not sure on how these components will look as well, even without the move to Vue3.
> 
>     * Dropping this would remove the need for types/index.d.ts and /index.ts files
> 

Exporting JS as TS shouldn't be an issue, but it's definitely something that we want to tackle soon after we stand up the initial version of the component library. My focus here is to limit the scope of changes by making incremental improvements. Layering multiple refactors increases complexity by a great deal and I'd rather avoid this if possible 😁 #5932

> Other points
> 
>     * `cd pkg/rancher-components && yarn lint` fails, it's assuming untyped vue components are typed/ts

Good eye 🙂. This will be resolved with #5932

>     * ` cd pkg/rancher-components && yarn run build:lib` errors with lots of
>       ```
>       Syntax Error: Thread Loader (Worker 1)
>       Failed to load config "standard" to extend from.
>       Referenced from: /home/richard/dev/github/rancher/dashboard-phil-ts/.eslintrc.js
>          at _normalizeObjectConfigDataBody.next (<anonymous>)
>          at _normalizeObjectConfigData.next (<anonymous>)
>       ```

This is new to me, I'll try to reproduce.. Can you provide some details about your environment?

cnotv commented on pkg/rancher-components/src/components/Card/Card.vue:
```diff
@@ -11,7 +12,8 @@ export default {
     buttonAction: {
       type:    Function,
-      default: () => {}
+      // TODO: is this the correct default? Originally () => { }
```
The original function is expected to do not return any value (falsy/undefined). I am not so sure if returning a value (truthy/true) may generate issues.

cnotv replied:
The original function is expected to do not return any value (falsy/undefined). I am not so sure if returning a value (truthy/true) may generate issues.

cnotv commented on pkg/rancher-components/src/components/Form/Checkbox/Checkbox.vue:
```diff
@@ -1,4 +1,5 @@
 <script>
+// TODO: Remove jquery dep
```
Should we maybe point to the issue number, so we know it's tracked?

cnotv replied:
Should we maybe point to the issue number, so we know it's tracked?

cnotv commented on pkg/rancher-components/src/components/Form/LabeledInput/index.ts:
```diff
@@ -0,0 +1 @@
+export { default as LabeledInput } from './LabeledInput.vue';
```
`LabeledInput` has now tests, please do not forget to move them.

cnotv commented on pkg/rancher-components/src/components/Form/index.ts:
```diff
@@ -0,0 +1,4 @@
+export * from './Checkbox';
```
If we are introducing barrel exports, we should probably inform it in the PR notes or maybe mention it in the documentation.
Could you also maybe explain why is this needed?

rak-phillip replied:
See my comment https://github.com/rancher/dashboard/pull/5504#discussion_r876381466

cnotv commented on pkg/rancher-components/src/components/Form/Checkbox/index.ts:
```diff
@@ -0,0 +1 @@
+export { default as Checkbox } from './Checkbox.vue';
```
I am not sure to understand the scope of creating an empty TS file with another export.

rak-phillip replied:
Doing something like this provides us with two benefits:

1. This provides us with a way to perform named exports without disrupting the requirement that Vue components be the default export. In essence, this allows us to `import { Checkbox, Cart, LabeledInput, etc.. } from '@rancher/components'` after we publish
2. It's an organizational technique to manage component exports close to the source. If we focus on Radio Buttons and form as an example, we can see that `RadioButton` & `RadioGroup` are exported via `pkg/rancher-components/src/components/Form/Radio/index.ts`. This export passed all the way up to `pkg/rancher-components/src/index.js`

An alternative might be to maintain a long list of exports at the project root. I didn't perceive that to be very nice to maintain, especially with the number of potential components that we could be publishing. This technique was influenced by other frameworks like Vuetify.

cnotv replied:
Ok this makes totally sense, to have some barrel imports. I was trying to replace the path to see if on fact was that the scope but I had issues about it.
Are you sure is going to work later as library? Also why is not working already with the existing export?

rak-phillip replied:
This works using a library. I use the library is my primary method of testing, it is the goal of this PR after all 😁

Anything we're doing here with these aliases and reaching directly into `pkg/rancher-components` for imports is just an intermediary step until we can get to a point where we can publish our first version of the component library

cnotv commented on pkg/rancher-components/src/mixins/labeled-form-element.js:
```diff
@@ -0,0 +1,139 @@
+    resizeHandler(e) {
+      // since the DD is positioned there is no way to 'inherit' the size of the input, this calcs the size of the parent and set the dd width if it is smaller. If not let it grow with the regular styles
+      this.$nextTick(() => {
+        const DD = $(this.$refs.select).find('ul.vs__dropdown-menu');
```
Could we link a ticket to this jQuery removal TODO?

cnotv commented on pkg/rancher-components/tsconfig.json:
```diff
@@ -0,0 +1,48 @@
+{
```
We should avoid creating duplicated configuration and import a default one.

rak-phillip replied:
I think it makes sense to do so, but I've run into issues with extending `tsconfig.default.json` because it defines `paths`.. I might be missing something, but I'm not coming across an elegant fix other than:

1. Refactor `tsconfig.default.json` so it doesn't define paths and allow individual components to define their own paths for aliases like `@shell` 
2. Understand that importing from `@shell` is a temporary measure until get into a position to publish a utils library and refactor at a later date

I'm open to other suggestions..

cnotv replied:
Not so sure which problem you faced, however I tried by replacing this content and everything seems working: build, dev, test, linting, IDE imports.

```json
{
  "extends": "../../tsconfig.default.json",
  "exclude": [
    "node_modules"
  ]
}
```

There's some issues when I run the `yarn build:lib` command, which I guess is what you were mentioning, that would need to exclude some paths. 

```
ERROR in /Users/gleo/projects/dashboard/node_modules/@types/express/index.d.ts(58,55):
58:55 Type 'Response<any, Record<string, any>>' does not satisfy the constraint 'ServerResponse'.
  Types of property 'req' are incompatible.
    Type 'Request<ParamsDictionary, any, any, ParsedQs, Record<string, any>> | undefined' is not assignable to type 'IncomingMessage'.
      Type 'undefined' is not assignable to type 'IncomingMessage'.
    56 |      * This is a built-in middleware function in Express. It serves static files and is based on serve-static.
    57 |      */
  > 58 |     var static: serveStatic.RequestHandlerConstructor<Response>;
       |                                                       ^
    59 | 
    60 |     /**
    61 |      * This is a built-in middleware function in Express. It parses incoming requests with urlencoded payloads and is based on body-parser.
```

rak-phillip replied:
That's correct, `build:lib` fails and that's the piece that needs to be working. I'll spend some time seeing if I can't come up with a solution, but I created #5996 in the event that this requires more effort..

cnotv commented on shell/nuxt.config.js:
```diff
@@ -610,7 +612,16 @@ export default function(dir, _appConfig) {
+      typeCheck: {
+        eslint: {
+          exclude: [
+            './pkg/rancher-components/*'
```
Why do we exclude this path?

rak-phillip replied:
I'd like to see rancher components to start making use of rules that make sense for a component framework (e.g. `vue/no-mutating-props`) so we can properly warn about unintended side effects.

From a previous conversation in this PR 

> I don't find the walking behavior to be desirable as I want rancher-components to make use of @vue/standard so we can start focusing on some improvements in our components. Mutating props is one thing that comes to mind as something that doesn't track as a good practice for a component library

cnotv commented on tsconfig.json:
```diff
@@ -3,5 +3,12 @@
+    "dist", 
+    "cypress", 
+    "shell/creators",
+    "pkg/rancher-components"
```
Why do we exclude this path?

rak-phillip replied:
I don't think this is needed.

cnotv reviewed (changes requested):
I have found some non-breaking missing files during the rebase.

Also while launching the app I see a lot of warnings related to the components, here one example:

```
client.js:196 ./pkg/rancher-components/src/components/Banner/index.tsModule Warning (from ./node_modules/eslint-loader/dist/cjs.js):Failed to load config "@vue/standard" to extend from.Referenced from: /Users/gleo/projects/dashboard/pkg/rancher-components/.eslintrc.js
```

rak-phillip commented:
> I have found some non-breaking missing files during the rebase.

I'll need some more info to better understand which files are missing..

> Also while launching the app I see a lot of warnings related to the components, here one example:
> 
> ```
> client.js:196 ./pkg/rancher-components/src/components/Banner/index.tsModule Warning (from ./node_modules/eslint-loader/dist/cjs.js):Failed to load config "@vue/standard" to extend from.Referenced from: /Users/gleo/projects/dashboard/pkg/rancher-components/.eslintrc.js
> ```

Thanks for the heads up. I can see this in my web console, that's where you're seeing it too, correct?

These warnings will go away once we publish the library and update our imports, but I'll see what we might be missing to resolve these warnings.

rak-phillip commented on pkg/rancher-components/jest.config.js:
```diff
@@ -0,0 +1,6 @@
+module.exports = {
```
I was originally looking into jest projects. I can't remember the issue I had, but I opted to keep jest close to components.

https://jestjs.io/docs/configuration#projects-arraystring--projectconfig

I'm not seeing any issue with deferring the running of unit tests to the project root, so we probably don't need this config or any jest dependencies in the components package.

rak-phillip commented on pkg/rancher-components/src/components/BadgeState/index.ts:
```diff
@@ -0,0 +1 @@
+export { default as BadgeState } from './BadgeState.vue';
```
Fixed

cnotv commented on pkg/rancher-components/.eslintrc.js:
```diff
@@ -0,0 +1,47 @@
+module.exports = {
```
I also have noticed that there's a lot of linting warnings, but I get them only if I run `yarn lint` in your path, but not if I run `yarn lint` in the root. We probably want to have it working in both the cases, so maybe you can configure the rules to be enabled for your path only? It my be verbose in a long run though.

rak-phillip replied:
These will be resolved once we refactor to typescript in #5932

We also have one component that is mutating props (`src/components/Form/Checkbox/Checkbox.vue`) and that will require refactoring as well. 

I think fixing these lint warnings are outside of the scope of this PR

cnotv commented:
> > I have found some non-breaking missing files during the rebase.
> 
> I'll need some more info to better understand which files are missing..

It is fine, you have already moved them 👍 

> **edit** I think this is fixed, can you make sure that you can run dev and build without any errors?

Sorry to be the bearer of bad news, but I got this error now:
<img width="591" alt="Screenshot 2022-05-19 at 12 36 41" src="https://user-images.githubusercontent.com/5009481/169274641-ca12e03b-6fda-45f0-a9b0-1ad833a489d8.png">

rak-phillip commented:
> Sorry to be the bearer of bad news, but I got this error now:
> <img alt="Screenshot 2022-05-19 at 12 36 41" width="591" src="https://user-images.githubusercontent.com/5009481/169274641-ca12e03b-6fda-45f0-a9b0-1ad833a489d8.png">

OK, I think I got this one fixed through the usage of yarn workspaces. I attempted multiple solutions to configure webpack with varying degrees of success, but all my testing revealed inconsistent results. Yarn workspaces appears to be a straightforward enough approach and hasn't failed for me yet, but I will continue testing just to be sure..

Can you try again to confirm that the issue was resolved? 

🗒️ With workspaces enabled, make sure to only `yarn install` from the project root. **Don't run `yarn install` under `pkg/rancher-components`. These dependencies should be handled via workspaces**.

rak-phillip commented on pkg/rancher-components/package.json:
```diff
@@ -0,0 +1,52 @@
+  ],
+  "types": "./types/index.d.ts",
+  "scripts": {
+    "build:lib": "vue-cli-service build --target lib --name @rancher/components src/main.ts",
```
I think adding `build:lib` to the root project is a good idea.

richard-cox reviewed (changes requested):
There's still some unresolved comments, I've updated some. There's quite a lot to track, at this point would it be quicker/easier to get us all on a call? Your shout.

- cd pkg/rancher-components --> yarn lint fails 
  - lots of @typescript-eslint/explicit-module-boundary-types
  - couldn't find a ticket for this
- cd pkg/rancher-components --> yarn run build:lib succeeds but contains warnings
- The bundle size is huge. 
  - Covered by https://github.com/rancher/dashboard/issues/5931
  - There's a line in our nuxt config which when uncommented uses the `webpack-bundle-analyzer` package to make a nice diagram. That's well worth trying (indepedently of nuxt) on the components lib 
- Minor styling issue
  - create project --> members --> resource quotas --> height of limit fields has grown (when compared with 2.6.5)

rak-phillip commented:
> There's still some unresolved comments, I've updated some. There's quite a lot to track, at this point would it be quicker/easier to get us all on a call? Your shout.
> 
>     * cd pkg/rancher-components --> yarn lint fails
>       
>       * lots of @typescript-eslint/explicit-module-boundary-types
>       * couldn't find a ticket for this

#5932 will cover this. I've added COA to mention that there should be no @typescript related lint warnings as aresult
 
>     * cd pkg/rancher-components --> yarn run build:lib succeeds but contains warnings

These are the same lint warnings as above. #5932 
 
>     * The bundle size is huge.
>       
>       * Covered by #5931
>       * There's a line in our nuxt config which when uncommented uses the `webpack-bundle-analyzer` package to make a nice diagram. That's well worth trying (indepedently of nuxt) on the components lib
> 
>     * Minor styling issue
>       
>       * create project --> members --> resource quotas --> height of limit fields has grown (when compared with 2.6.5)

I'll investigate to see what's changed

rak-phillip commented on pkg/rancher-components/src/components/BadgeState/BadgeState.vue:
```diff
@@ -87,6 +87,7 @@ export default {
   }
 </style>
 <style lang="scss">
+  // TODO: Investigate why this is here.. I don't think that styles for sortable table should belong here
```
Added #6005 to track this

rak-phillip commented:
> Minor styling issue
> 
>     * create project --> members --> resource quotas --> height of limit fields has grown (when compared with 2.6.5)

This is an existing issue in `v2.6-head` (`0e2fdc8`), unrelated to this change. Here's what a new instance of Dashboard looks like this morning

![image](https://user-images.githubusercontent.com/835961/169566158-9b7c5710-0aba-4424-9987-6ce86b07556d.png)

rak-phillip commented on pkg/rancher-components/src/main.ts:
```diff
@@ -0,0 +1,19 @@
+// import * as components from './index'
+
+// const ComponentLibrary = {
+//   install (Vue) {
+//     for (const componentName in components) {
+//       const component = components[componentName]
+
+//       Vue.component(component.name, component)
+//     }
+//   }
+// }
+
+// export default ComponentLibrary
+
+// if (typeof window !== 'undefined' && window.Vue) {
+//   window.Vue.use(ComponentLibrary)
+// }
```
Will delete this for the moment and we can pick it up later if there's demand

richard-cox reviewed (approved):
With the file removed and merge conflict resolved, LGTM
